File: leetcode/8-Valid-Parentheses/Valid-Parentheses.cs
//Runtime: 76 ms, faster than 54.99% of C# online submissions for Valid Parentheses.
//Memory Usage: 20 MB, less than 39.25% of C# online submissions for Valid Parentheses.

public class Solution {
    public bool IsValid(string s) {
            Stack<Char> st = new Stack<char>();
            if (s.Length == 0)
            {
                return true;
            }
            if (s.Length%2 == 1)
            {
                return false;
            }
            if(s[0]=='}'|| s[0] == ']'|| s[0] == ')')
            {
                return false;
            }
            for (int i = 0; i < s.Length; i++)
            {
                if (s[i] == '[' || s[i] == '{' || s[i] == '(')
                {
                    st.Push(s[i]);
                }
                if (s[i] == ')' && st.Pop() != '(')
                {
                    return false;
                }
                if (s[i] == ']' && st.Pop() != '[')
                {
                    return false;
                }
                if (s[i] == '}' && st.Pop() != '{')
                {
                    return false;
                }
            }
            if (st.TryPop(out char a))
            {
                return false;
            }
            return true;
    }
}

File: leetcode/8-Valid-Parentheses/Valid-Parentheses.go
//Runtime: 0 ms, faster than 100.00% of Go online submissions for Valid Parentheses.
//Memory Usage: 6 MB, less than 11.96% of Go online submissions for Valid Parentheses.

func isValid(s string) bool {
	if len(s) == 0{
		return true
	}
	if len(s)%2 == 1{
		return false
	}
	if s[0]==')'||s[0]=='}'||s[0]==']'{
		return false
	}
	var st = ""
	for i:=0;i<len(s);i++{
		if s[i]=='('||s[i]=='{'||s[i]=='['{
			st = st + string(s[i])
			continue
		}
		if s[i] == ']' && st[len(st)-1] != '['{
			return false
		}
		if s[i] == ')' && st[len(st)-1] != '('{
			return false
		}
		if s[i] == '}' && st[len(st)-1] != '{'{
			return false
		}
		st = st[:len(st)-1]
	}
	if len(st) != 0{
		return false
	}
	return true
}